### DISCUSSION TITLE
"illegal image mode" when trying to subtract a png from another by using putalpha

### DISCUSSION THREAD

**Discussion by Nicolas-Gth** [python-pillow/Pillow#7295] [Q&A] [answered]
This is my code

```py
reflexionBackground = Image.new('RGBA', image.size, (255, 0, 0, 0))
reflexion = Image.open("./assets/ProgressBar/Reflexion.png")
reflexionBackground.alpha_composite(reflexion, (50, 55))
reflexionBackground.save("reflexionBackground.png")

substractShapeBackGround = Image.new('RGBA', image.size, (255, 0, 0, 0))
substractShape = Image.open("./assets/ProgressBar/SubstractShape.png")
substractShapeBackGround.alpha_composite(substractShape, (fillXCoord, 0))
substractShapeBackGround.save("substractShapeBackGround.png")

substractShapeBackGround.convert('L').resize(reflexionBackground.size)
reflexionBackground.putalpha(substractShapeBackGround)
```
reflexionBackground.png :
![reflexionBackground](https://github.com/python-pillow/Pillow/assets/78755503/80976f50-699c-4900-be97-fd8d0fb13825)
substractShapeBackground :
![substractShapeBackground](https://github.com/python-pillow/Pillow/assets/78755503/7fd0005f-5808-4644-97bb-52dd61fa4ad1)

I'd like to substract the black shape from reflexionBackground.png

**Comment by radarhere** [comment, marked as answer]
I think you've correctly understood that the `alpha` argument to `putalpha()` "[can either be an “L” or “1” image](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.putalpha)"

However, what I think you've missed is that `convert()` "[Returns a converted copy of this image](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert)". A copy. It does not modify the original. The same goes for [`resize()`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.resize).

So I would suggest just changing
```python
substractShapeBackGround.convert('L').resize(reflexionBackground.size)
```
to
```python
substractShapeBackGround = substractShapeBackGround.convert('L').resize(reflexionBackground.size)
```

**Reply by Nicolas-Gth** [reply]
Oh ok, thanks ! However I still have a problem :

If I do `reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))`
and `substractShapeBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))`
I get
![reflexionBackgroundFinal](https://github.com/python-pillow/Pillow/assets/78755503/f3f440f3-f6ed-4b3f-8268-a47e80260538)

But if I do `reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))`
and `substractShapeBackground = Image.new('RGBA', image.size, (255, 255, 255, 0))`
I get
![reflexionBackgroundFinal](https://github.com/python-pillow/Pillow/assets/78755503/dc09ce14-837d-4d7d-9f71-bb9e81082d10)

I don't understand why changing the RGB values makes a difference if they're set to transparent, and I would like to have my reflection with transparency on a fully transparent background.

**Reply by radarhere** [reply]
So far, you've attached reflexionBackground.png and substractShapeBackground.png.
Could you attach Reflexion.png and SubstractShape.png?

Also, your code never defines `image`.

**Reply by Nicolas-Gth** [reply]
Sorry, I don't understand what you mean by "Could you attach Reflexion.png and SubstractShape.png?".


This was only a snippet of my code, here's the full code where image is defined :
```py
from PIL import Image, ImageFont, ImageDraw


def drawProgressBar(value, maxValue, fillPercentage, startColor, endColor):

	image = Image.open("./assets/ProgressBar/BackGround.png")
	ctx = ImageDraw.Draw(image)

	#position 
	xCoord = 18
	yCoord = 24
	length = 1762
	height = 120

        # calculates progress bar coordinates
	fillXCoord = int(length*fillPercentage) + xCoord - 24

	# rectangle with gradient
	for i in range(length):
		r = int(startColor[0] + (endColor[0] - startColor[0]) * i / length)
		g = int(startColor[1] + (endColor[1] - startColor[1]) * i / length)
		b = int(startColor[2] + (endColor[2] - startColor[2]) * i / length)
		ctx.line((xCoord + i, yCoord, xCoord + i, yCoord + height), fill=(r,g,b))

	# overlay layers
	shadow = Image.open("./assets/ProgressBar/Shadow.png")
	image.alpha_composite(shadow, (5, 44))

	# subtracts the shape from the reflection
	reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))
	reflexion = Image.open("./assets/ProgressBar/Reflexion.png")
	reflexionBackground.alpha_composite(reflexion, (50, 55))
	reflexionBackground.save("reflexionBackground.png")

	substractShapeBackground = Image.new('RGBA', image.size, (255, 0, 0, 0))
	substractShape = Image.open("./assets/ProgressBar/SubstractShape.png")
	substractShapeBackground.alpha_composite(substractShape, (fillXCoord, 0))
	substractShapeBackground.save("substractShapeBackground.png")
	
	substractShapeBackground = substractShapeBackground.convert('L')
	substractShapeBackground.save("substractShapeBackground.png")
	reflexionBackground.putalpha(substractShapeBackground)
	reflexionBackground.save("reflexionBackgroundFinal.png")
	image.alpha_composite(reflexionBackground, (0, 0))

	# slider layer
	slider = Image.open("./assets/ProgressBar/Slider.png")
	image.alpha_composite(slider, (fillXCoord, 23))

	frame = Image.open("./assets/ProgressBar/Frame.png")
	image.alpha_composite(frame, (0, 0))

	#texte
	W, H = image.size
	font = ImageFont.truetype("./assets/Fonts/BAHNSCHRIFT.TTF", size=75)

	text = f"{value} / {maxValue}"
	_, _, w, h = ctx.textbbox((0, 0), text, font=font)
	ctx.text(((W-w)/2, (H-h)/2), text, font=font, fill=(255,255,255))

	ctx.text((textXCoord,textYCoord), text, fill=(255,255,255), font=font)

	return image

# test
drawnImage = drawProgressBar(600, 1000, 0.6, (0, 43, 156), (0, 170, 255))
drawnImage.save("ProgressBar.png")
```
[assets.zip](https://github.com/python-pillow/Pillow/files/12123683/assets.zip)

**Reply by radarhere** [reply]
Could you upload the images that your code is opening here, in a comment, so that we can access them? That way, we can run your code the same way that you can.

**Reply by Nicolas-Gth** [reply]
I added a zip file with the assets

My goal is to get this :
![ProgressBar](https://github.com/python-pillow/Pillow/assets/78755503/6a61a237-4625-4dbb-ac8d-7bf19a3b5869)

**Reply by radarhere** [reply]
> If I do `reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))` and `substractShapeBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))` I get ![reflexionBackgroundFinal](https://user-images.githubusercontent.com/78755503/255163977-f3f440f3-f6ed-4b3f-8268-a47e80260538.png)
> 
> But if I do `reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))` and `substractShapeBackground = Image.new('RGBA', image.size, (255, 255, 255, 0))` I get ![reflexionBackgroundFinal](https://user-images.githubusercontent.com/78755503/255164226-dc09ce14-837d-4d7d-9f71-bb9e81082d10.png)
> 
> I don't understand why changing the RGB values makes a difference if they're set to transparent, and I would like to have my reflection with transparency on a fully transparent background.

`substractShapeBackGround.convert('L')` is converting an RGBA image to L, a greyscale mode. To do that, Pillow simply ignores the transparency at that point, and only uses the RGB values. So that is why different RGB values produces different results.

If your intention was to get the alpha channel, I would suggest `substractShapeBackGround.getchannel("A")` instead. Using that, both of your scenarios here produce the same `reflexionBackground` in the end.

**Reply by Nicolas-Gth** [reply]
Mmh, maybe I wasn't clear when explaining what I am trying to obtain.

So after executing this code :
```py
reflexionBackground = Image.new('RGBA', image.size, (0, 0, 0, 0))
reflexion = Image.open("./assets/ProgressBar/Reflexion.png")
reflexionBackground.alpha_composite(reflexion, (50, 55))
reflexionBackground.save("reflexionBackground.png")

substractShapeBackground = Image.new('RGBA', image.size, (255, 255, 255, 0))
substractShape = Image.open("./assets/ProgressBar/SubstractShape.png")
substractShapeBackground.alpha_composite(substractShape, (fillXCoord, 0))

substractShapeBackground = substractShapeBackground.convert('L')
substractShapeBackground.save("substractShapeBackground.png")
```
 I have these two images :
 
![reflexionBackground](https://github.com/python-pillow/Pillow/assets/78755503/75f58d88-8fc1-4449-b387-6b7412597810)

![substractShapeBackGround](https://github.com/python-pillow/Pillow/assets/78755503/65617530-604e-4087-8481-83727c3612f9)

What I would like to obtain :
![reflexionBackgroundFinal](https://github.com/python-pillow/Pillow/assets/78755503/9436ddad-6e40-460a-8f90-5822a3bcb858)

What
```py
reflexionBackground.putalpha(substractShapeBackground)
reflexionBackground.save("reflexionBackgroundFinal.png")
```
gives me :
![reflexionBackgroundFinal](https://github.com/python-pillow/Pillow/assets/78755503/5195dc49-4284-41a2-9cad-17ade6d823e0)

**Reply by radarhere** [reply]
Ok, then, how about this? The following code
```python
output = Image.new("RGBA", reflexionBackground.size)
output.paste(reflexionBackground, mask=substractShapeBackground)
output.save("current.png")
```
creates
![current](https://github.com/python-pillow/Pillow/assets/3112309/d3e054d3-16ee-45aa-8ff6-7f0247f753e4)

**Reply by Nicolas-Gth** [reply]
Oh, yes it's working ! Thank you so much
